Oryx.Cognite/src/Handler.fs

```diff
@@ -63,14 +63,21 @@ module Handler =
         | ResponseError error -> raise error
         | Panic err -> raise err
 
-    /// Use the given handler token provider for the request.
-    let withTokenProvider<'TResult, 'TNext, 'TError> (tokenProvider: Func<CancellationToken, Task<string>>) (handler: HttpHandler<HttpResponseMessage, 'TNext, 'TResult, 'TError>) =
-        let provider ct = task {
-            let! token = tokenProvider.Invoke ct
-            return Option.ofObj token
+    /// Use the given handler token provider for the request. Adapts a C# function to F#.
+    let withTokenRenewer<'TResult, 'TNext, 'TError> (tokenProvider: Func<CancellationToken, Task<string>>) (handler: HttpHandler<HttpResponseMessage, 'TNext, 'TResult, 'TError>) =
+        let renewer ct = task {
+            try
+                let! token = tokenProvider.Invoke ct
+                match Option.ofObj token with
+                | Some token ->
+                    return Ok token
+                | None ->
+                    return Panic (NullReferenceException "No token received.") |> Error
```

Comment:
I don't `NullReferenceException` should be used like this. I agree the runtime would throw in in this case anyway, but still, NullReferenceException usually means null is not handled, but here it's handled

Seems there is no common practice for manual null checks: https://stackoverflow.com/questions/34163767/what-exception-should-i-throw-for-an-unexpected-null-value :/

Reply:
Sorry, I had a missing commit there. I fully agree and I changed it to ArgumentNullException which is slightly better. But you might not like that either because it's for an argument and not a return value.